### ISSUE TITLE
gitlab CI build of vcpkg

### ISSUE THREAD

**Issue by pwm1234** [microsoft/vcpkg#5776]
I am trying to set up a gitlab continuous integration build of vcpkg. I have a build script that bootstraps vcpkg and then installs and exports the packages or ports that I have specifiied. But the problem is that if/when I add another package, then the entire vcpkg install and export is repeated. This is because gitlab cleans everything that is not part of the git repo before it does the CI build. And I do not think this clean step is avoidable. 

While this is the right behavior for CI builds of most repos, this makes it hard to use vcpkg, where the build is done in the source tree and I have some heavyweight packages that I am installing (boost, opencv with cuda, qt5, ...) this takes hours for vcpkg to build, which is unavoidable the first time. But when I later decide there is another package that I want vcpkg to install, then the whole build is done again. 

Is there a better way to do a CI build of vcpkg? What is the recommended way to use vcpkg in a CI build? (Perhaps my biggest problem is that I am using gitlab CI and there is no checkbox to skip the git clean step?)

Thanks!

**Comment by pwm1234** [comment]
I discovered the [gitlab CI cache][1]. I think proper use of the cache is key for this issue. Does anybody have a ``.gitlab-ci.yml`` file that builds vcpkg?

  [1]: https://docs.gitlab.com/ee/ci/yaml/README.html#cache

**Comment by pwm1234** [comment]
Thank you! Yes, this is what I was looking for. Another question for you if I may: from where should I get a base docker image for windows development? (Docker images/container running windows is still a mystery to me. I appreciate any pointers.)

**Comment by pwm1234** [comment]
Thank you! Yes, this is what I was looking for. Another question for you if I may: from where should I get a base docker image for windows development? (Docker images/container running windows is still a mystery to me. I appreciate any pointers.)

**Comment by pwm1234** [comment]
Thank you! Yes, this is what I was looking for. Another question for you if I may: from where should I get a base docker image for windows development? (Docker images/container running windows is still a mystery to me. I appreciate any pointers.)

**Comment by MoritzMaxeiner** [comment]
> from where should I get a base docker image for windows development

You'll have to build it yourself. You can start by taking a look [at ours](https://git.imp.fu-berlin.de/bioroboticslab/robofish/docker/blob/master/windows/base/Dockerfile) and remove what you don't need.
With regards to setting up a docker gitlab-runner for windows, that currently requires [some manual hacks](https://gitlab.com/andrewleech/gitlab-runner/) until [upstream has proper support](https://gitlab.com/gitlab-org/gitlab-runner/issues/2609).
Basically instead of the normal `gitlab-runner.exe` you'll need the one provided by Andrew (and obviously a Windows server box with the "Container" feature enabled and [native docker installed](https://docs.docker.com/install/windows/docker-ee/#update-docker-ee)). Other than that you can mostly follow the normal procedure for setting up a docker executor.

**Comment by NancyLi1013** [comment]
Hi @pwm1234 thanks for posting this issue. 
If you have no other questions about this, please help close this issue.
@MoritzMaxeiner thanks for your detail answers to this question.
